src/focus-history.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {timer} from './timer';
+
+
+/**
+ * FocusHistory keeps track of recent focused elements. This history can be
+ * purged using `purgeBefore` method.
+ */
+export class FocusHistory {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const {!Window} */
+    this.win = win;
+
+    /** @private @const {!Array<!{el: !Element, time: time}>} */
+    this.history_ = [];
+
+    /** @private @const {function(!Event)} */
+    this.focusCapture_ = e => {
+      if (e.target) {
+        this.pushFocus_(e.target);
+      }
+    };
+    this.blurCapture_ = e => {
+      timer.delay(() => {
+        this.pushFocus_(this.win.document.activeElement);
+      }, 500);
+    };
+    this.win.document.addEventListener('focus', this.focusCapture_, true);
+    this.win.addEventListener('blur', this.blurCapture_);
+  }
+
+  /** @private For testing. */
+  cleanup_() {
+    this.win.document.removeEventListener('focus', this.focusCapture_, true);
+    this.win.removeEventListener('blur', this.blurCapture_);
+  }
+
+  /**
+   * @param {!Element} element
+   * @private
+   */
+  pushFocus_(element) {
+    if (this.history_.length == 0 ||
+            this.history_[this.history_.length - 1].el != element) {
+      this.history_.push({el: element, time: timer.now()});
+    } else {
+      this.history_[this.history_.length - 1].time = timer.now();
+    }
+  }
+
+  /**
+   * Returns the element that was focused last.
+   * @return {!Element}
+   */
+  getLast() {
+    if (this.history_.length == 0) {
+      return null;
+    }
+    return this.history_[this.history_.length - 1].el;
+  }
+
+  /**
+   * Removes elements from the history older than the specified time.
+   * @param {time} time
+   */
+  purgeBefore(time) {
+    let index = this.history_.length - 1;
+    for (let i = 0; i < this.history_.length; i++) {
+      if (this.history_[i].time >= time) {
+        index = i - 1;
+        break;
+      }
+    }
+    if (index != -1) {
+      this.history_.splice(0, index + 1);
+    }
+  }
+
+  /**
+   * Returns `true` if the specified element contains any of the elements in
+   * the history.
+   * @param {!Element} element
+   * @return {boolean}
+   */
+  hasDescendantsOf(element) {
```

Comment:
Maybe I'm missing something, but for the case we actually care about the element is the actual iframe, no?